flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/allocator/SlotSharingSlotAllocator.java

```diff
@@ -150,21 +175,31 @@ public Optional<JobSchedulingPlan> determineParallelismAndCalculateAssignment(
      * distributed over the remaining groups.
      */
     private static Map<SlotSharingGroupId, Integer> determineSlotsPerSharingGroup(
-            JobInformation jobInformation, int freeSlots) {
+            JobInformation jobInformation,
+            int freeSlots,
+            int minRequiredSlots,
+            Map<SlotSharingGroupId, Integer> minSlotsPerSlotSharingGroup) {
+
         int numUnassignedSlots = freeSlots;
         int numUnassignedSlotSharingGroups = jobInformation.getSlotSharingGroups().size();
+        int numMinSlotsRequiredByRemainingGroups = minRequiredSlots;
 
         final Map<SlotSharingGroupId, Integer> slotSharingGroupParallelism = new HashMap<>();
 
         for (Map.Entry<SlotSharingGroupId, Integer> slotSharingGroup :
                 sortSlotSharingGroupsByDesiredParallelism(jobInformation)) {
+            final int minParallelism = minSlotsPerSlotSharingGroup.get(slotSharingGroup.getKey());
+
             final int groupParallelism =
-                    Math.min(
-                            slotSharingGroup.getValue(),
-                            numUnassignedSlots / numUnassignedSlotSharingGroups);
+                    minParallelism
+                            + Math.min(
+                                    slotSharingGroup.getValue() - minParallelism,
+                                    (numUnassignedSlots - numMinSlotsRequiredByRemainingGroups)
+                                            / numUnassignedSlotSharingGroups);
```

Comment:
It would be great to have a comment here on this calculation, why we have 2 ways to compute this (and take the minimum) because it's not very obvious at first glance